### DISCUSSION TITLE
Upcoming changes in Wöhler analysis module

### DISCUSSION THREAD

**Discussion by johannes-mueller** [boschresearch/pylife#104] [General]
## Synopsis

One of the pyLife's popular features is the Wöhler analysis. It has been initially developed as a master theses and then been in use now for a couple of years. With time comes experience and we realize that a couple of changes will be necessary in order to make sure that it is all according to the state of the art. In this discussion page there will be the possibility to discuss the details before anything is rolled out.

### Shutdown of the Bayesian Wöhler analyzer

That was one of the exciting things of the master thesis, as it was a new idea to use Bayesian optimization to estimate a Wöhler curve from SN-data. In the meantime we collected experience with it and it turns out, that there are quite a few cases in which the output is not ideal. Moreover the dependencies to pymc turn out to be quite troublesome. That's why we with heavy heart decided to shut it down. That means that although the code will remain in the code base for people to study it, it will raise an exception if someone tries to import it.

We invite everyone to improve the module and we will happily reactivate it if it has been improved in a convincing way.

### New parameter `fatigue_limit`

This is quite a simple improvement. Users expressed the wish to manually set the limit between finite and infinte zone, the so called fatigue limit. This is now possible.

### Use only finite data for calculation of the slope `k_1`

The standard DIN 50100:2022-12 suggests to only use the data of the finite zone i.e. pure fracture levels to calculate the slope `k_1`. So far pyLife does not adhere to that but uses all the fracture points. This will be changed. We are aware that that will lead to subtle changes in the analysis results, but adhering to standards is important in engineering. 

This change will open an interesting edge case. What if there is only one pure fracture level? So far we could still calculate a slope because we had the mixed levels. With only one load level however, it is not possible to calculate a slope. It would still be interesting to do the analysis because the endurance limit in load direction `SD` and the scatter `TS` can still be estimated. However, we cannot simply set the slope value to `inf` which will mean a flat Wöhler curve, because we know about the pure fracture level and thus must suppose that the Wöhler curve is not flat.

Our current stance is to set `k_1 = NaN` and `TN = 1` in that case although we are aware of the troubles this will cause downstream in the damage calculation. But in any way you might put it, we know that there is a slope, but we can't calculate it. So a succeeding damage calculation would be actually a scam.

## Time line

There will be a beta release for version `2.1.1` beginning of August 2024. The final release of `2.1.1` will not happen before beginning of September. That means that all of you have a month to contribute to the discussion. Finally, when the discussion is settled we will do the release with the changes.

## How to test

You can install the current betas using the following command.

```
pip install --pre "pylife>=2.1.1b1,<2.2.0"
```

**Comment by uphill-bob** [comment]
Hi Johannes,
do you intend to update the dependencies to Numpy > 2.0 also?
Currently, that combination does not run into a dependency error, but usage of numpy.asfarray yields an error when executing the module.
Otherwise, adapting the dependencies accordingly would be helpful.

**Reply by johannes-mueller** [reply]
The `numpy.asfarray` issue is resolved in the current `master` branch as of d3ac921. There will be a pyLife 2.1.1 release today, so that the fix will be finally roled out.

**Comment by johannes-mueller** [comment]
The above proposed changes have been made and release in [pylife-2.1.1](https://github.com/boschresearch/pylife/releases/tag/2.1.1)
